tracks/cpp/exercises/binary-search-tree/mentoring.md

```diff
@@ -0,0 +1,20 @@
+### Reasonable solution ###
+There is an example [in the github repo](https://github.com/exercism/cpp/blob/master/exercises/binary-search-tree/example.h) for this track.
```

Comment:
The `example.h` is _not_ the same as the reasonable solution. Remove this and copy the solution here.

Reply:
Hi. I didn't quite understand you... Do you mean the file doesn't contain the solution, or that the solution should be embedded?

I noticed most commentary has the solution embedded in the file, but I think this will be to robust (the file is 227 lines), and the github file can get updates...

Reply:
> that the solution should be embedded?

The solution should be embedded. The example files in each track are for passing the CI, not necessarily a reasonable solution. The issue that keeping them in sync is a lot of overhead for both maintainers and the `website-copy` team.

Reply:
> I noticed most commentary has the solution embedded in the file, but I think this will be to robust (the file is 227 lines)

Personally I don't think that's an issue, but I _understand_ your point. The only other reasonable solution would be to duplicate the file in `cpp` track and rename it, with a contributing note that that file has to be the reasonable solution.

Pinging @exercism/cpp.

Reply:
If we duplicated the file in the `cpp` track, would the new file be sent to students? I thought that exercism sends all files except for the `example`s.

While 200+ lines is a lot to paste into a `README`, an embedded copy would help keep the reasonable solution in sync with its discussion. I'm concerned that the contents of `example.h` could drift over time, leaving the Talking Points out of date.

@patricksjackson may have a more educated opinion - he's been more involved in reviewing this exercise.

Reply:
Well this is interesting. I was very much under the impression that the example solutions in the github repo were used by mentors if they needed to look at a reasonable solution, and have been operating under that assumption for a few _years_ now. I'm fairly sure that assumption is true for all the maintainers in both the C and C++ tracks.

I assumed that track maintainers and track mentors would work together to keep reasonable solutions in the track, not a copy inside of a markdown file _where it would never be tested against new versions of the tests_. So I fail to see why the README file would be more up to date than the example file. It makes a lot more sense to me to have the example be the 'reasonable' solution.

All that said, I've never had time to mentor so I've only been maintaining the track. Obviously we'll defer to you here.

Reply:
The reasonable solution might not be the one in the repo because:
1. There might be a few reasonable solutions
2. The solution in the repo might be better/worse than the one we expect students to use.
Specifically, in our case, we can approve solutions where the constructor gets a ref to const, or overloaded on lvalue/rvalue refs, but the solution in the repo uses universal refs.

Reply:
What @elyashiv is correct! Apart from that, they are _not_ kept in sync. 

@patricksjackson I've heard that before here and there. Sometimes on slack a mentor links one and usually this happens when the example.xxx file is sub-optimal. People just infer it, but it's not said anywhere.

The example.xxx files are really just there for CI to pass, so that we know the exercise _can_ be solved. But because most maintainers are not part of the website-copy group, we can no longer assume the reasonable solution equals the example.xxx file.